test/src/files.test.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2024 Holger Dal Mogensen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as assert from 'assert'
+import * as vscode from 'vscode'
+import { activate, getTestDocUri, sleep } from './util'
+
+suite('Source file manipulation', () => {
+  const doc1Uri = getTestDocUri('src/Temp1.flix')
+  const doc1Content = `def foo(): Unit = bar()`
+
+  const doc2Uri = getTestDocUri('src/Temp2.flix')
+  const doc2Content = `def bar(): Unit = ()`
+
+  async function addFile(uri: vscode.Uri, contents: string) {
+    await vscode.workspace.fs.writeFile(uri, Buffer.from(contents))
+  }
+
+  suiteSetup(async () => {
+    await activate()
+    await addFile(doc1Uri, doc1Content)
+  })
+  suiteTeardown(async () => {
+    try {
+      await vscode.workspace.fs.delete(doc1Uri)
+    } catch {
+      // Ignore
+    }
+  })
+  teardown(async () => {
+    try {
+      await vscode.workspace.fs.delete(doc2Uri)
+    } catch {
+      // Ignore
+    }
+  })
+
+  async function docIsAdded() {
+    // Wait for the diagnostics to be reported
+    // TODO: Do this in a smarter way
+    await sleep(1000)
```

Comment:
Perhaps we could simply wait in a loop without a timeout? It seems more robust-- if we have to wait at least.

Reply:
I don't believe we have any way of actually knowing if the diagnostics have arrived. Before they're ready, vscode reports an empty array, indistinguishable from zero diagnostics being present.

Reply:
Yes. I know, My proposal was just to try (i.e. see if we have the diagnostics) every 100ms and timeout after 5000ms (i.e. fail the test).

Reply:
As opposed to waiting 1000ms which may be both too little and too much.

Reply:
But in one of the tests we are asserting that none are present

Reply:
Ooooh. Hmm!

Reply:
But lets not do that- because its brittle. When testing add/remove can we just test that hover works correctly?

Reply:
Same problem - only appears after some delay

Reply:
I believe the problem is that VS Code doesn't wait for the server to process the addition event. As I see it we have 3 possibilities:
1. Add some command like "flix.awaitFileAdded" or something like that, which only resolves once the compiler responds.
2. Make this test less of an integration test and instead just test the server's handler directly.
3. Add some delay like we are doing now. I'm guessing that 1s should be more than enough.